### ISSUE TITLE
能不能通过领域数据对模型进行微调

### ISSUE THREAD

**Issue by Crazycatter** [ymcui/Chinese-LLaMA-Alpaca#146]
能不能通过领域数据对模型进行微调

**Comment by airaria** [comment]
有数据的话当然可以。

**Comment by delltower** [comment]
> 有数据的话当然可以。

您好 如何用领域数据进行二次预训练呢

**Comment by airaria** [comment]
只是换数据而已，方法上和普通的预训练没有区别。

**Comment by delltower** [comment]
> Collaborator

谢谢！请问有训练框架推荐吗

**Comment by zhihui-shao** [comment]
同问如何用领域数据训练

**Comment by airaria** [comment]
预训练的话参考 https://github.com/huggingface/transformers/tree/main/examples/pytorch/language-modeling 里的run_clm.py
精调的话是按[Stanford Alpaca](https://github.com/tatsu-lab/stanford_alpaca)来做的。

**Comment by delltower** [comment]
> 
您好 readme上说加入了20G中文语料进行继续训练，想参考机器配置和训练时间，我想在垂直领域继续训练，做个参考

**Comment by delltower** [comment]
> 先训练的题目参考https://github.com/huggingface/transformers/tree/main/examples/pytorch/language-modeling里的run_clm.py 精调的题目是按[Stanford Alpaca](https://github.com/tatsu-lab/stanford_alpaca)来做的。


您好readme上说加入了20G中文资料继续训练，想参考机器配置和训练时间，我想在垂领域继续训练，做个参考

**Comment by ruanshudong** [comment]
微调的数据比较好理解, 通常是input+labels方式输入即可, 预训练的时候, 数据是怎么组织? input labels怎么来的?

**Comment by airaria** [comment]
> > 先训练的题目参考[https://github.com/huggingface/transformers/tree/main/examples/pytorch/language-modeling里的run_clm.py](https://github.com/huggingface/transformers/tree/main/examples/pytorch/language-modeling%E9%87%8C%E7%9A%84run_clm.py) 精调的题目是按[Stanford Alpaca](https://github.com/tatsu-lab/stanford_alpaca)来做的。
> 
> 您好readme上说加入了20G中文资料继续训练，想参考机器配置和训练时间，我想在垂领域继续训练，做个参考

使用LoRA的方式预训练7B模型，在20G语料上，八卡A100-48G训练1轮大约120小时；当然机器更多会更快。

**Comment by airaria** [comment]
> 



> 微调的数据比较好理解, 通常是input+labels方式输入即可, 预训练的时候, 数据是怎么组织? input labels怎么来的?

CLM方式预训练，inputs向左偏移一位就是对应的labels

**Comment by ruanshudong** [comment]
@airaria 多谢! 这里transformer里面有可以参考的代码么?

**Comment by airaria** [comment]
> @airaria 多谢! 这里transformer里面有可以参考的代码么?

你注意看下，LlamaForCausalLM中forward方法下有相应的处理代码可参考

**Comment by ruanshudong** [comment]
@airaria 多谢!

**Comment by ruanshudong** [comment]
词表合并这里, 我理解是将: 遍历llama模型的tokenizer, 将它的词表, 添加到sentencepiece生成的词表后面, 这个逻辑没问题吧? 这其中 score设置为0, 会有问题么?

**Comment by airaria** [comment]
> 词表合并这里, 我理解是将: 遍历llama模型的tokenizer, 将它的词表, 添加到sentencepiece生成的词表后面, 这个逻辑没问题吧? 这其中 score设置为0, 会有问题么?

到目前为止，我们还没发现这么做有什么影响

**Comment by ruanshudong** [comment]
收到, 我也试一试. 我还有一个问题, 预训练过程中, 文本如何断句输入, 按行来做input么?每次输入一行?

**Comment by ruanshudong** [comment]
@airaria 预训练过程中为啥内存会用到80G(GPU还没有开始使用), 内存占用过大, 被kill了, 这可能是什么原因?如何解决?

**Comment by yuanzhiyong1999** [comment]
> 

请问解决了吗？

**Comment by XYCsony** [comment]
> > 
> 
> 请问解决了吗？

还没解决，你也遇到同样问题了嘛？

**Comment by yuanzhiyong1999** [comment]
我看你的代码base_model选择llama，peftmodel选择chinesellama，这样的方式能行的通吗？



在 2023-04-27 10:16:48，"XYCsony" ***@***.***> 写道：

请问解决了吗？

还没解决，你也遇到同样问题了嘛？

**Comment by WuNein** [comment]
> 作者，你好，我在用我的数据集上继续指令微调的时候遇到了loss为0的问题，想问一下作者或者其他人有没遇到同样的问题，问题、代码、训练参数下所示： <img alt="bd68f1ad58dd8d71f99d57c54d2c453" width="341" src="https://user-images.githubusercontent.com/13764151/234632815-61275ee9-f445-43d9-a425-4b40c4ddf163.png"> 代码：`
> 
> ```
> tokenizer = LlamaTokenizer.from_pretrained(“ziqingyang/chinese-alpaca-lora-7b”, padding_side="left")
> base_model = LlamaForCausalLM.from_pretrained(
>     "decapoda-research/llama-7b-hf",
>     # load_in_8bit=True,
>     load_in_8bit=False,
>     torch_dtype=torch.float16,
>     device_map=device_map,
> )
> base_model.resize_token_embeddings(len(tokenizer))
> # config = PeftConfig.from_pretrained(lora_model)
> # model = get_peft_model(base_model, config)
> # 改成加载别人训练好的模型
> model = PeftModel.from_pretrained(base_model, “ziqingyang/chinese-alpaca-lora-7b”) # lora_model :"ziqingyang/chinese-alpaca-lora-7b")`
> ```
> 
> 训练参数： batch_size: int = 64, micro_batch_size: int = 2, num_epochs: int = 1, learning_rate: float = 2e-5, cutoff_len: int = 512, val_set_size: int = 64,


说起来LLaMA的默认padding不在left？
隔壁代码也是要手动指定
https://github.com/tloen/alpaca-lora/blob/8bb8579e403dc78e37fe81ffbb253c413007323f/finetune.py#L124
用了RoPE模型的padding到底应该在那一侧呢？

**Comment by airaria** [comment]
> > 作者，你好，我在用我的数据集上继续指令微调的时候遇到了loss为0的问题，想问一下作者或者其他人有没遇到同样的问题，问题、代码、训练参数下所示： <img alt="bd68f1ad58dd8d71f99d57c54d2c453" width="341" src="https://user-images.githubusercontent.com/13764151/234632815-61275ee9-f445-43d9-a425-4b40c4ddf163.png"> 代码：`
> > ```
> > tokenizer = LlamaTokenizer.from_pretrained(“ziqingyang/chinese-alpaca-lora-7b”, padding_side="left")
> > base_model = LlamaForCausalLM.from_pretrained(
> >     "decapoda-research/llama-7b-hf",
> >     # load_in_8bit=True,
> >     load_in_8bit=False,
> >     torch_dtype=torch.float16,
> >     device_map=device_map,
> > )
> > base_model.resize_token_embeddings(len(tokenizer))
> > # config = PeftConfig.from_pretrained(lora_model)
> > # model = get_peft_model(base_model, config)
> > # 改成加载别人训练好的模型
> > model = PeftModel.from_pretrained(base_model, “ziqingyang/chinese-alpaca-lora-7b”) # lora_model :"ziqingyang/chinese-alpaca-lora-7b")`
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 训练参数： batch_size: int = 64, micro_batch_size: int = 2, num_epochs: int = 1, learning_rate: float = 2e-5, cutoff_len: int = 512, val_set_size: int = 64,
> 
> 说起来LLaMA的默认padding不在left？ 隔壁代码也是要手动指定 https://github.com/tloen/alpaca-lora/blob/8bb8579e403dc78e37fe81ffbb253c413007323f/finetune.py#L124 用了RoPE模型的padding到底应该在那一侧呢？

我们实验用的都是right padding

**Comment by han508** [comment]
请问loss=0，解决了吗，我把llama模型的梯度冻结，只微调lora，训练正常，可是当把llama的embed层的梯度设置为True，然后同时训练embd与lora，loss只有第一个有值，然后一直是0.

**Comment by han508** [comment]
代码：import torch
model = LlamaForCausalLM.from_pretrained('decapoda-research/llama-7b-hf',device_map='auto',cache_dir='./cache/',load_in_8bit = True)
model.resize_token_embeddings(len(merge_tokenizer))
from transformers import TrainingArguments, Trainer, DataCollatorForLanguageModeling
trainArgs = TrainingArguments(
  output_dir= '../ckps',
    do_train=True,
    per_device_train_batch_size=4,
    gradient_accumulation_steps=4,
    evaluation_strategy="steps",
    save_strategy="steps",
    save_steps=1000,
    eval_steps=100,
    logging_steps=1,
    warmup_steps=100,
    num_train_epochs=2,
    learning_rate=3e-4,
    
    push_to_hub=False,
    load_best_model_at_end=True,

    report_to="wandb"
)
model = prepare_model_for_int8_training(model)
lora_config = LoraConfig(
    r=8,
    lora_alpha=16,
    target_modules=["q_proj", "v_proj"],
    lora_dropout=0.05,
    bias="none",
    task_type="CAUSAL_LM",
)
model = get_peft_model(model, lora_config)

for name, param in model.named_parameters():
    if name =='base_model.model.model.embed_tokens.weight':
        param.requires_grad_(True)
    print(name, param.device, "requires_grad:", param.requires_grad)

trainer = Trainer(
    model=model,
    args=trainArgs,
    train_dataset=input_data,
    eval_dataset=input_data,
    data_collator=DataCollatorForLanguageModeling(merge_tokenizer, mlm=False),
)
# old_collator = trainer.data_collator
# trainer.data_collator = lambda data: dict(old_collator(data))
# model.config.use_cache = False
old_state_dict = model.state_dict
model.state_dict = (
    lambda self, *_, **__: get_peft_model_state_dict(self, old_state_dict())
).__get__(model, type(model))

**Comment by airaria** [comment]
> 请问loss=0，解决了吗，我把llama模型的梯度冻结，只微调lora，训练正常，可是当把llama的embed层的梯度设置为True，然后同时训练embd与lora，loss只有第一个有值，然后一直是0.

是使用了int8 training？
int8 training下的确有这个问题，等peft更新吧

**Comment by han508** [comment]
感谢回复，是在int8下。当我把int8关掉，然后通过peft lora 微调的时候，报错提示tensor 在gpu和cpu上，打印模型权重的device后，发现lora的权重全在cpu上，其余代码与上面的我发的一致，请问这怎么解决呢？

**Comment by han508** [comment]
新的peft 发布了https://github.com/huggingface/peft/releases/tag/v0.3.0

**Comment by kevinuserdd** [comment]
> > 词表合并这里, 我理解是将: 遍历llama模型的tokenizer, 将它的词表, 添加到sentencepiece生成的词表后面, 这个逻辑没问题吧? 这其中 score设置为0, 会有问题么?
> 
> 到目前为止，我们还没发现这么做有什么影响

score为什么要设置为0，这样训练的权重怎么体现？

**Comment by guozhiyao** [comment]
请问能否开源fine tune时的代码呢？想知道增加了哪些可训练参数。

**Comment by airaria** [comment]
项目还在持续完成中；增加的可训练参数可参见adapter_config.json内容：包括embed_tokens, lm_head, attention中的qkvo和三个mlp矩阵的lora

**Comment by airaria** [comment]
> 请问能否开源fine tune时的代码呢？想知道增加了哪些可训练参数。

项目还在持续完成中；增加的可训练参数可参见adapter_config.json内容：包括embed_tokens, lm_head, attention中的qkvo和三个mlp矩阵的lora

**Comment by guozhiyao** [comment]
> > 请问能否开源fine tune时的代码呢？想知道增加了哪些可训练参数。
> 
> 项目还在持续完善；增加的可训练参数可参见adapter_config.json内容：包括embed_tokens, lm_head, attention中的qkvo和三个mlp矩阵的lora

请问下，pretrain和fine tune是用两套不同的lora？还是fine tune是基于pretrain的lora继续训练呢？

**Comment by airaria** [comment]
> > > 请问能否开源fine tune时的代码呢？想知道增加了哪些可训练参数。
> > 
> > 
> > 项目还在持续完善；增加的可训练参数可参见adapter_config.json内容：包括embed_tokens, lm_head, attention中的qkvo和三个mlp矩阵的lora
> 
> 我看预训练脚本里已经配置了这些参数：
> 
> ```
> lora_trainable="q_proj,v_proj,k_proj,o_proj,gate_proj,down_proj,up_proj"
> modules_to_save="embed_tokens,lm_head"
> ```
> 
> 还想和ft没有很大区别？ 此外，pretrain和fine tune是用两套不同的lora？还是fine tune是基于pretrain的lora继续训练呢？

ft里也是这样配置的。

对于非plus模型，pretrain的lora是基于fine tune的lora继续训练的
对于plus模型，pretrain和fine tune用的两套lora

**Comment by guozhiyao** [comment]
> > > > 请问能否开源fine tune时的代码呢？想知道增加了哪些可训练参数。
> > > 
> > > 
> > > 项目还在持续完善；增加的可训练参数可参见adapter_config.json内容：包括embed_tokens, lm_head, attention中的qkvo和三个mlp矩阵的lora
> > 
> > 
> > 我看预训练脚本里已经配置了这些参数：
> > ```
> > lora_trainable="q_proj,v_proj,k_proj,o_proj,gate_proj,down_proj,up_proj"
> > modules_to_save="embed_tokens,lm_head"
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 还想和ft没有很大区别？ 此外，pretrain和fine tune是用两套不同的lora？还是fine tune是基于pretrain的lora继续训练呢？
> 
> ft里也是这样配置的。
> 
> 对于非plus模型，pretrain的lora是基于fine tune的lora继续训练的 对于plus模型，pretrain和fine tune用的两套lora

所以ft和pretrain的lora参数是一样的咯？因为我看这里这么说，所以有点疑问
<img width="954" alt="image" src="https://github.com/ymcui/Chinese-LLaMA-Alpaca/assets/21999339/c9b187e0-c87d-4ba3-959e-b3c9efb21d71">

此外，对于非plusmo模型，是不是写错了？是fine turn基于pretrain的？

**Comment by airaria** [comment]
> > > > > 请问能否开源fine tune时的代码呢？想知道增加了哪些可训练参数。
> > > > 
> > > > 
> > > > 项目还在持续完善；增加的可训练参数可参见adapter_config.json内容：包括embed_tokens, lm_head, attention中的qkvo和三个mlp矩阵的lora
> > > 
> > > 
> > > 我看预训练脚本里已经配置了这些参数：
> > > ```
> > > lora_trainable="q_proj,v_proj,k_proj,o_proj,gate_proj,down_proj,up_proj"
> > > modules_to_save="embed_tokens,lm_head"
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > 还想和ft没有很大区别？ 此外，pretrain和fine tune是用两套不同的lora？还是fine tune是基于pretrain的lora继续训练呢？
> > 
> > 
> > ft里也是这样配置的。
> > 对于非plus模型，pretrain的lora是基于fine tune的lora继续训练的 对于plus模型，pretrain和fine tune用的两套lora
> 
> 所以ft和pretrain的lora参数是一样的咯？因为我看这里这么说，所以有点疑问 <img alt="image" width="954" src="https://user-images.githubusercontent.com/21999339/237281204-c9b187e0-c87d-4ba3-959e-b3c9efb21d71.png">
> 
> 此外，对于非plusmo模型，是不是写错了？是fine turn基于pretrain的？


因为实验方案一直在迭代，所以不同模型(7B/13B/plus)在训练细节上有所差别，细节可以参看[技术报告](https://arxiv.org/abs/2304.08177)，较详细描述了7B/13B非plus版的训练方案

**Comment by guozhiyao** [comment]
> > > > > > 请问能否开源fine tune时的代码呢？想知道增加了哪些可训练参数。
> > > > > 
> > > > > 
> > > > > 项目还在持续完善；增加的可训练参数可参见adapter_config.json内容：包括embed_tokens, lm_head, attention中的qkvo和三个mlp矩阵的lora
> > > > 
> > > > 
> > > > 我看预训练脚本里已经配置了这些参数：
> > > > ```
> > > > lora_trainable="q_proj,v_proj,k_proj,o_proj,gate_proj,down_proj,up_proj"
> > > > modules_to_save="embed_tokens,lm_head"
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > 还想和ft没有很大区别？ 此外，pretrain和fine tune是用两套不同的lora？还是fine tune是基于pretrain的lora继续训练呢？
> > > 
> > > 
> > > ft里也是这样配置的。
> > > 对于非plus模型，pretrain的lora是基于fine tune的lora继续训练的 对于plus模型，pretrain和fine tune用的两套lora
> > 
> > 
> > 所以ft和pretrain的lora参数是一样的咯？因为我看这里这么说，所以有点疑问 <img alt="image" width="954" src="https://user-images.githubusercontent.com/21999339/237281204-c9b187e0-c87d-4ba3-959e-b3c9efb21d71.png">
> > 此外，对于非plusmo模型，是不是写错了？是fine turn基于pretrain的？
> 
> 因为实验方案一直在迭代，所以不同模型(7B/13B/plus)在训练细节上有所差别，细节可以参看[技术报告](https://arxiv.org/abs/2304.08177)，较详细描述了7B/13B非plus版的训练方案

好的，十分感谢。
